[Royden Lim Yi Ren] ip

roydenlyr opened the pull request.

roydenlyr committed:
Add initial skeletal version

roydenlyr committed:
Add echo feature, exit command, and style output message

roydenlyr committed:
feat(list): add list feature

- Implement array to store user input
- Implement list command to display content of array sequentially
- Implement sanitizeTask function to handle typographical errors by user
- Implement switch-case for command query

roydenlyr committed:
feat(task status): add status for each task

- Implement Task class to store task detail
- Implement mark and unmark command
- Implement extractTrailingNumber function to update desired task

roydenlyr committed:
feat(persona): personalise chatbot responses

- Replaced default Duke responses with ShinChan style lines
- Centralised message in persona class
- Update bye, list, mark, unmark and default response tone

roydenlyr committed:
refactor(task): add formatTask method

- Move repeated getStatusIcon + getDescription logic into Task class
- Simplifies list, mark and unmark outputs

roydenlyr committed:
refactor: replace string concat with StringBuilder for task listing

roydenlyr committed:
refactor(io): reuse single Scanner instance

- Move Scanner init outside while loop
- Prevent unnecessary Scanner object creation on System.in
- Improved resource handling and code efficiency

roydenlyr committed:
fix(input): handle invalid user input

- Add bounds checking for task indices in mark/unmark
- Prevent crash when input number missing or out of range
- Prevent empty input for add command

roydenlyr committed:
fix(mark): add missing parenthesis in conditional so task prints correctly

Jeweljace commented on src/main/java/Persona.java:
```diff
@@ -0,0 +1,86 @@
+public class Persona {
```
I like the personality of your chatbot.

Jeweljace commented on src/main/java/Persona.java:
```diff
@@ -0,0 +1,86 @@
+        return ("""
+                To add or to ignore? Hmmm... Fine, I'll add it this time. TSK!
+                """);
+    }
```
I like how the output messages have their own separate methods which makes the main code way neater.

Jeweljace commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+    public String printTask() {
+        return (getStatusIcon() + " " + getDescription());
+    }
+
```
Should this empty line be removed?

Jeweljace commented on src/main/java/Shinchan.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+        int taskIndex = extractTrailingNumber(line) - 1;
+        if (taskIndex < 0) return;
```
Should the if statement have curly brackets?

Jeweljace reviewed (commented):
Overall I like the code:)

Amanda-HUANG88 commented on src/main/java/Persona.java:
Clear Separation of Concerns! This code effectively encapsulates all user-facing text and personality traits within the Persona class.

Amanda-HUANG88 commented on src/main/java/Persona.java:
The customized chatbot is highly creative.

Amanda-HUANG88 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,40 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    private static int numOfTasks = 0;
```
Regarding the static variable numOfTasks, since it only increments in the constructor, it can't handle task deletion. If a task were to be removed, the total count may be incorrect?

Amanda-HUANG88 commented on src/main/java/Shinchan.java:
```diff
@@ -0,0 +1,99 @@
+                "added: " + tasks[Task.getNumOfTasks() - 1].getDescription());
+    }
+
+    public static void updateTaskStatus(Task[] tasks, String line, boolean isDone, Persona persona) {
+        if (Task.getNumOfTasks() == 0) {
+            persona.listIntro();
+            return;
+        }
+        int taskIndex = extractTrailingNumber(line) - 1;
+        if (taskIndex < 0) return;
+        if (taskIndex > Task.getNumOfTasks()) {
+            printMessage("Hello? That task number is out of bounds!");
+            return;
+        }
+        tasks[taskIndex].setDone(isDone);
+        printMessage((isDone ? persona.markIntro() : persona.unmarkIntro()) + tasks[taskIndex].printTask());
+    }
```
Great refactoring! The codes are neat and readable.

Amanda-HUANG88 reviewed (commented):
Overall, look good to me !

roydenlyr committed:
refactor(command): replace string-based switch with enums

- Introduced Command enum to represent valid user commands
- Replaced string-based parsing with enum-based switch

roydenlyr committed:
refactor(Task): implement toString override

- Override toString() to print formatted task
- Remove printTask() from Task class

roydenlyr committed:
refactor(command): extract showList method and rename sanitizeMessage to extractCommand

roydenlyr committed:
feat(task): add Todo, Deadline, Event commands

- Implemented corresponding classes:
 * Todo extends Task
 * Deadline extends Todo
 * Event extends Deadline
- Update command handler to construct and display these tasks

roydenlyr committed:
test: update input.txt and EXPECTED.txt for automated testing

roydenlyr committed:
refactor(task): extract date formatting and  unify content parsing

- Introduced formateDate() to encapsulate date formatting logic used by
  deadline and event tasks
- Renamed extractContent() to extractContents() and refactored it to
  return a String[] array shared by addTodo, addDeadline and addEvent
- Renamed Personal#addTaskDescription to addTask and overloaded it:
  * addTask() for default task addition
  * addTask(String) for Todo, Deadline, and Event message summary

roydenlyr committed:
feat(exception): Add exception handling for invalid commands

- Implemented try-catch blocks to capture invalid commands
- Added custom exception messages for empty or incomplete task inputs
- Ensured user-friendly feedback instead of abupt program termination

roydenlyr committed:
Merge branch 'branch-Level-5' into master

roydenlyr committed:
refactor: restructure packages for commands, exceptions, tasks and ui

roydenlyr committed:
Merge branch 'branch-A-Packages' into main

roydenlyr committed:
refactor: replace fixed-size array with ArrayList

roydenlyr committed:
feat(command): add 'delete' command to remove specific tasks

- Implemented new command 'delete' for removing tasks by-index
- Added deleteTask() method in Persona class to provide user feedback
- Updated command enum to recognise DELETE

roydenlyr committed:
feat(storage): persist todo list to text file

- Added Datamanager class to handle file creation, reading, and writing
- Implemented saving of todo list into a text file
- Implemented loading of todo list from file on program start
- Ensured file is auto-created if missing in directory
- Centralised file I/O logic

roydenlyr committed:
Merge branch 'branch-Level-6'

roydenlyr committed:
Merge branch 'branch-Level-7'

roydenlyr committed:
feat(delete): update storage to reflect deleted tasks

okkhoy commented on src/main/java/shinchan/Shinchan.java:
```diff
@@ -0,0 +1,220 @@
+
+        Scanner input = new Scanner(System.in);
+
+            while(true) {
```
can extract the contents to another method for better SLAP

okkhoy commented on src/main/java/shinchan/Shinchan.java:
```diff
@@ -0,0 +1,220 @@
+                        printMessage("Invalid command");
+                        break;
+                    }
+                } catch (TaskNumberOutOfBoundException | MarkMissingItemNumberException | TaskMissingDateException
```
not a good practice to OR the exceptions.
better to have individual catch clauses for each type of exception, so that you can do customized exception handling

okkhoy commented on src/main/java/shinchan/Shinchan.java:
```diff
@@ -0,0 +1,220 @@
+        printMessage(msg.toString());
+    }
+
+    public static void updateTaskStatus(ArrayList<Task> taskList, String line, boolean isDone, Persona persona)
```
all non-trivial classe/methods need javadoc comments

okkhoy commented on src/main/java/shinchan/Shinchan.java:
```diff
@@ -0,0 +1,220 @@
+            return -1;
+        }
+    }
+}
```
Main class needs serious refacotring and reorganization for better code organization and maintainability

okkhoy commented on src/main/java/shinchan/data/DataManager.java:
```diff
@@ -0,0 +1,141 @@
+        return taskList;
+    }
+
+    private ArrayList<Task> parse(ArrayList<String> dataItems) {
```
all non-trivial methods/classes need header comments.

okkhoy commented on src/main/java/shinchan/data/DataManager.java:
```diff
@@ -0,0 +1,141 @@
+                allTasks.add(event);
+                break;
+            default:
+                System.out.println("Unknown task encountered. Skipping");
```
can this be moved to the UI somehow?

okkhoy reviewed (commented):
- Generally, the code is neat
- I don't see much header comments; all non-trivial methods/classes need javadoc comments.
- SLAP in the code needs improvement

roydenlyr committed:
refactor(tasklist): Add TaskList class to perform tasks operations

roydenlyr committed:
refactor(Commands): Add Parse class to extract commands from user input

roydenlyr committed:
refactor(exception): introduce ShinChanException base class

- Added abstract class ShinChanException as parent for all custom
  exceptions
- Refactored existing custom exceptions to extend ShinChanException
- Declared Persona.printMessage() as a static method for consistent
  usage

roydenlyr committed:
feat(command): add 'find' to search tasks by keyword

- Implemented new 'find' command to filter tasks containing a keyword

roydenlyr committed:
Merge branch 'branch-Level-9'

roydenlyr committed:
feat(command): support date formatting and add 'find date' command

- Implemented date input format for Deadline and Event tasks
- Added new 'find date' command to list tasks matching a given date

roydenlyr committed:
Merge branch 'branch-Level-8'

roydenlyr committed:
refactor(TaskList): slim TaskList class to model-only

roydenlyr committed:
refactor(commands): added command interface and IndexParser class

- Add command interface to group all individual commands class in the
  future
- Added IndexParser class to keep code clean and readable

roydenlyr committed:
refactor(delete): added DeleteCommand class to handle delete commands

roydenlyr committed:
refactor(command): added Mark, Unmark, and List command classes

roydenlyr committed:
refactor(core): adopt Command pattern and decouple TaskList from UI/I/O

- Introduce command classes:
  - AddTodoCommand, AddDeadlineCommand, AddEventCommand
  - FindWordCommand, FindDateCommand
  - ByeCommand (signals exit), EmptyCommand (no-op prompt), ErrorCommand (invalid/unsupported input)
- Remove legacy parser in commands/ and replace with dedicated parser package:
  - Add shinchan.parsers.Parser (returns Command objects)
  - Add FlagTokenizer (parse /description, /by, /from, /to)
  - Add DateParser (strict "yyyy-MM-dd HHmm")
  - Move IndexParser to shinchan.parsers
- Model refinements:
  - Task: add occursOn(LocalDateTime) hook
  - Deadline/Event: implement occursOn(...) (deadline equality; event inclusive range)
  - TaskList: make domain-only (no UI/storage), expose find(...) and asUnmodifiableList()
- UI:
  - Add TaskRenderer for numbered list rendering
  - Update Persona messages to work with command outputs
- App entry point:
  - Update Shinchan.java to use new parsers.Parser and Command execution flow

Rationale: improves separation of concerns (logic vs model vs UI), follows SLAP/guard clauses,
and increases testability and maintainability of parsing and command execution.

BREAKING CHANGE: command parsing now routed through shinchan.parsers.Parser; the old
commands.Parser class is deleted.

roydenlyr committed:
refactor(storage): drop legacy format and persist canonical pipe records

- DataManager now reads/writes only canonical records:
	T|done|desc
	D|done|desc|yyyy-MM-dd HHmm
	E|done|desc|from|to
- Remove legacy [T]{ } UI-string parsing
- Use Path/UTF-8/atomic write
- Keep static writeToFile/appendToFile for backward compatibility (for
  now)

roydenlyr committed:
rename(exception): renamed MarkMissingItemNumberException to MissingNumberItemException

roydenlyr committed:
refactor(parser, commands): tighten input validation and error UX

- parsers
  - DateParser: normalise whitespace; enforce strict 'yyyy-MM-dd HHmm';
    add clearer messages
  - IndexParser: accept only non-negative integers
- commands
  - AddEventCommand: validate /from <= /to;
- exceptions: add InvalidDateException, InvalidItemNumberException;
- ui(persona): changed chatbot's tone

roydenlyr committed:
Merge branch 'branch-Refactor-TaskList'

roydenlyr committed:
feat(commands): add Javadoc to most files except exceptions and Persona

roydenlyr committed:
Merge branch 'branch-A-JavaDoc'

roydenlyr committed:
docs(readme): update README to serve as User Guide

roydenlyr committed:
Merge pull request #1 from roydenlyr/branch-Update-ReadMe

docs(readme): update README to serve as User Guide

roydenlyr committed:
docs(readme): Reformat commmand table

roydenlyr committed:
docs(readme): Reformat 'Saving Data' bullets

roydenlyr committed:
Update test cases and EXPECTED.txt

roydenlyr committed:
Add more help instructions for TaskMissingDataException
